core/src/main/java/com/netflix/conductor/core/execution/tasks/Wait.java

```diff
@@ -12,29 +12,120 @@
+
     public Wait() {
         super(TASK_TYPE_WAIT);
     }
 
+    private static Duration parseDuration(String text) {
+        Matcher m =
+                Pattern.compile(
+                                "\\s*(?:(\\d+)\\s*(?:days?|d))?"
+                                        + "\\s*(?:(\\d+)\\s*(?:hours?|hrs?|h))?"
+                                        + "\\s*(?:(\\d+)\\s*(?:minutes?|mins?|m))?"
+                                        + "\\s*(?:(\\d+)\\s*(?:seconds?|secs?|s))?"
+                                        + "\\s*",
+                                Pattern.CASE_INSENSITIVE)
+                        .matcher(text);
+        if (!m.matches()) throw new IllegalArgumentException("Not valid duration: " + text);
+
+        int days = (m.start(1) == -1 ? 0 : Integer.parseInt(m.group(1)));
+        int hours = (m.start(2) == -1 ? 0 : Integer.parseInt(m.group(2)));
+        int mins = (m.start(3) == -1 ? 0 : Integer.parseInt(m.group(3)));
+        int secs = (m.start(4) == -1 ? 0 : Integer.parseInt(m.group(4)));
+        return Duration.ofSeconds((days * 86400) + (hours * 60L + mins) * 60L + secs);
+    }
+
+    private static Date parseDate(String date) throws ParseException {
+        return DateUtils.parseDate(date, patterns);
+    }
+
     @Override
     public void start(WorkflowModel workflow, TaskModel task, WorkflowExecutor workflowExecutor) {
+
+        String duration =
+                Optional.ofNullable(task.getInputData().get(DURATION_INPUT)).orElse("").toString();
+        String until =
+                Optional.ofNullable(task.getInputData().get(UNTIL_INPUT)).orElse("").toString();
+
+        if (Strings.isNotBlank(duration) && Strings.isNotBlank(until)) {
+            task.setReasonForIncompletion(
+                    "Both 'duration' and 'until' specified. Please provide only one input");
+            task.setStatus(FAILED);
+            return;
+        }
+
+        if (Strings.isNotBlank(duration)) {
+
+            Duration timeDuration = parseDuration(duration);
+            task.getOutputData()
+                    .put(TIMEOUT, System.currentTimeMillis() + (timeDuration.getSeconds() * 1000));
+            long seconds = timeDuration.getSeconds();
+            task.setCallbackAfterSeconds(seconds);
+        } else if (Strings.isNotBlank(until)) {
+            try {
+                Date expiryDate = parseDate(until);
+                System.out.println("expiryDate: " + expiryDate);
+                long timeInMS = expiryDate.getTime();
+                long now = System.currentTimeMillis();
+                long seconds = (timeInMS - now) / 1000;
+                task.getOutputData().put(TIMEOUT, timeInMS);
+                task.setCallbackAfterSeconds(seconds);
+
+            } catch (ParseException parseException) {
+                task.setReasonForIncompletion(
+                        "Invalid/Unsupported Wait Until format.  Provided: " + until);
+                task.setStatus(FAILED);
```

Comment:
I think this could also be `FAILED_WITH_TERMINAL_ERROR`.

Reply:
done.